### DISCUSSION TITLE
개발 시나리오 및 세부 기능 구상

### DISCUSSION THREAD

**Discussion by jopopcorn** [euphony-io/check-in-speaker#1] [Ideas]
@euphony-io/euphoria 

앱의 목적과 기능을 확실히 추리기 위해 다음과 같은 사항들을 논의해야 합니다. 의견 자유롭게 남겨주세요! 😄

</br>

- 개발 시나리오 정하기
  - 앱을 어떤 식으로 사용할 건지?
  - 앱의 간단한 내용, 무슨 기능을 하는 앱인지?

</br>

- 기능 요구사항
  - 위의 개발 시나리오를 만족하기 위해 어떤 기능이 필요한가?
  - [기능 구현을 위해 필요한 기술, 아키텍처, 컨벤션 (여기서 자세히 이야기 나눕시다!)](https://github.com/euphony-io/wave-in-speaker/discussions/3)
  - 어떤 데이터를 음파로 생성할 건지? (백신 접종 여부, 개인 안심 번호, 시군구)
  - 빠른 수신을 위해 위 데이터를 어떻게 압축시킬건지?

</br>

- 플로우 차트
  - 구체적 화면 구성과 요소들을 정의
  - 화면과 시나리오를 고려해 플로우 차트 짜기

**Comment by psh2849** [comment]
저희 기능적인 부분에 대해서 생각해봤는데, 기존 안심번호는 숫자 4개와 한글 1자리로 구성된다고 합니다.
하지만, 저희 Euphony 프로젝트 같은 경우 현재 한글을 지원하고 있지 않기 때문에 전송 중 제대로 된 데이터 값이 아닌 다른 값이 전송될 것 같습니다. 따라서, 한글 대신 영어를 이용한다던지 해야 할 것 같습니다!

그리고.. 사용자의 안심번호를 저장할 데이터베이스가 있으면 좋을 것 같은데 현재 사용중인 안심번호는 이 안심번호를 이용해서 기존 번호를 역추적할 수 있기 때문에 데이터베이스에 저장할 때는 암호화를 해서 저장한다고 합니다! 

참고했던 자료 공유할게요!

> 안심번호 정보 

https://news.mt.co.kr/mtview.php?no=2021021818062145320

https://codefor.kr/wiki/%EA%B0%9C%EC%9D%B8%EC%95%88%EC%8B%AC%EB%B2%88%ED%98%B8

**Reply by jopopcorn** [reply]
오오 그렇군요 한글도 되긴 하는데 정확성을 위해 우선은 영어를 이용하는 게 좋겠어요.
데이터베이스에 암호화해서 저장하는 부분은 리스너 팀 분들께 공유해드리면 좋을 것 같네요!

**Comment by jopopcorn** [comment]
@euphony-io/euphoria 
기존의 QR 체크인 앱을 확인해보고 디자인을 하다보니 생각난 부분인데, 기존 앱들은 시군구 데이터 없이 그냥 안심번호만 보내주고 있더라고요.
만약 체크인을 한 뒤에 확진이 되면 그때 시군구 정보를 수집하지 않을까 싶습니다.
따라서 시군구 정보는 빼고 음파로 안심번호만 보내면 될 것 같은데, 다른 분들 생각은 어떤지 의견 묻고 싶습니다!

**Reply by psh2849** [reply]
궁금한게 시군구 정보가 사용자가 거주하는 주소인가요 아니면 저희가 QR 코드를 찍어야하는 가게들의 주소인가요??

**Reply by jopopcorn** [reply]
사용자의 시군구 정보입니다!

**Reply by psh2849** [reply]
그러면 처음에 전화번호 입력할때 같이 입력받는게 좋을까요??

**Reply by jopopcorn** [reply]
앗 제가 말한 의미는 확진되지 않는 이상 시군구 데이터는 없어도 돼서 수집하지 않아도 될 것 같다는 말씀입니다!
`(즉, 확진되면 코로나 확진자 관리 측에서 확진자에게 시군구 정보를 직접 물음)`

따라서 유저에게 전화번호 입력받기 -> 입력 받은 전화번호를 안심번호로 변환 & 내부에 저장 -> 음파 데이터 생성 하면 될 것 같습니다!

**Reply by psh2849** [reply]
아하 넵 알겠습니다!

**Reply by HACore** [reply]
나중에 기존 QR체크인 상용앱들과의 확장성 면에서도 시군구 정보는 빼고 전송하는게 더 좋을거같네요!

**Comment by HACore** [comment]
@euphony-io/euphoria
#5 에서 `'음성 데이터로 체크인하기 위해 볼륨을 크게 키워주세요'와 같은 안내 문구`를 추가한 부분을 보고 추가로 생각해본 기능이 있습니다.

보통 바코드나 큐알코드를 인식시키기 위해 앱에서 해당 화면을 띄우면 인식률을 높이기위해 화면이 자동으로 밝아지는 기능이 포함된 경우가 있는데, 저희 앱도 통신을 위해 사용자가 매번 볼륨을 조절하는 것은 번거로우니 통신 시작 시에 자동으로 일정 이상의 볼륨으로 세팅되도록 구현하는 것은 어떨까요?

혹시나 구현이 가능한지 찾아보니 `AudioManager`를 이용하면 될 것 같습니다! (https://15051015.tistory.com/93)

**Reply by psh2849** [reply]
좋은 아이디어인 것 같습니다!
말씀해주신대로 통신 시작 버튼을 눌렀을 때 적당한 볼륨으로 조절 시키면 좋을 것 같습니다!

혹여나 고려할 만한 점은 실수로 음악 듣다가 버튼을 누르는 경우에 귀 아플수도 있고..

`AudioManager`를 한번도 안써봐서 한번 의도대로 작동하는지 확인해볼게요! 감사합니다

**Reply by jopopcorn** [reply]
저도 좋은 아이디어라고 생각해요! 👍👍
일일이 볼륨 조절하기는 번거로우니 체크인 시작 버튼을 눌렀을 때는 소리를 크게 하고, 종료 시 원래 소리 크기로 돌려놓으면 좋을 것 같습니다!

아마 성호님이 고려하신 부분은 `Audio Focus` 처리를 통해 해결할 수 있을 것 같습니다!
버튼을 눌렀을 때 듣고 있는 음악은 중단되고, 음파는 재생되는 방식으로 구현해야 할 것 같습니다.
이것 또한 `AudioManager`에서 포커싱을 제어할 수 있는 것 같아서 이부분도 같이 작업하면 좋지 않을까 싶네요!

**Reply by psh2849** [reply]
`Audio Focus`도 한번 찾아봐야겠네요 ㅎㅎ 감사합니다!

**Comment by psh2849** [comment]
![image](https://user-images.githubusercontent.com/42669772/135883216-eed788dc-db9f-42b1-99b2-9e636cc2d030.png)

근데 저희 여기서 '?' 부분에 추가적인 기능이 있나요??

**Reply by jopopcorn** [reply]
앗 저 버튼을 클릭했을 때 개인안심번호에 관한 설명을 띄우려고 했는데 부가적인 기능이라 의논하는 걸 깜빡했네요ㅋㅋㅋㅋ
지금이라도 추가하는 게 좋을까요?!

**Reply by psh2849** [reply]
ㅋㅋㅋㅋㅋ 시간 나실 때 추가해주시면 감사하겠습니다!

**Reply by jopopcorn** [reply]
> ㅋㅋㅋㅋㅋ 시간 나실 때 추가해주시면 감사하겠습니다!

넵넵 리드미 올리기 전에 추가해야겠네요 ㅎㅎ 감사해요!
